### PR TITLE
Ddpb 808 edit client details

### PR DESCRIPTION
**Pull request by elvisciotti**

### CODE REVIEW

File: src/AppBundle/Controller/Pa/IndexController.php [hunk 1]
```diff
@@ -48,6 +49,40 @@ public function dashboardAction(Request $request)
...
+     * Client edit page
+     * Report is only associated to one client, and it's needed for back link routing,
+     * so it's retrieved with the report with a single API call
+     *
+     * @Route("/client/{clientId}/edit", name="pa_client_edit")
+     * @Template
+     */
+    public function clientEditAction(Request $request, $clientId)
+    {
+        /** @var $client EntityDir\Client */
```
Comment (slizzio) [inline src/AppBundle/Controller/Pa/IndexController.php]: ACL needed here (see comment on story and possible use of a new Voter?

File: src/AppBundle/Entity/Client.php [hunk 2]
```diff
@@ -15,23 +15,45 @@ class Client
...
+     * @JMS\Groups({"edit", "pa-edit"})
      *
+     * Lay
      * @Assert\NotBlank( message="client.firstname.notBlank" )
      * @Assert\Length(min=2, minMessage= "client.firstname.minMessage", max=50, maxMessage= "client.firstname.maxMessage")
      *
+     * PA
+     * @Assert\NotBlank( message="client.firstname.notBlank", groups={"pa-client"})
+     * @Assert\Length(min=2, minMessage= "client.firstname.minMessage", max=50, maxMessage= "client.firstname.maxMessage", groups={"pa-client"})
+     *
```
Comment (slizzio) [inline src/AppBundle/Entity/Client.php]: Wont the two sets of Assertions conflict with each other?

File: src/AppBundle/Entity/Client.php [hunk 3]
```diff
@@ -15,23 +15,45 @@ class Client
...
+     * @JMS\Type("string")
+     * @JMS\Groups({"edit", "pa-edit"})
+     *
+     * Lay
+     * @Assert\NotBlank( message="client.lastname.notBlank" )
+     * @Assert\Length(min = 2, minMessage= "client.lastname.minMessage", max=50, maxMessage= "client.lastname.maxMessage")
+     *
+     * PA
+     * @Assert\NotBlank( message="client.lastname.notBlank", groups={"pa-client"})
+     * @Assert\Length(min = 2, minMessage= "client.lastname.minMessage", max=50, maxMessage= "client.lastname.maxMessage", groups={"pa-client"})
```
Comment (slizzio) [inline src/AppBundle/Entity/Client.php]: Same here (multiple assertions ok?)

File: src/AppBundle/Entity/Client.php [hunk 4]
```diff
@@ -47,22 +69,19 @@ class Client
...
 
     /**
-     * @var Odr\Odr
+     * @JMS\Type("AppBundle\Entity\Report\Report")
      *
-     * @JMS\Type("AppBundle\Entity\Odr\Odr")
+     * @var array
      */
-    private $odr;
+    private $reportCurrent;
```
Comment (slizzio) [inline src/AppBundle/Entity/Client.php]: consider $currentReport?

File: src/AppBundle/Entity/Client.php [hunk 5]
```diff
@@ -97,42 +116,58 @@ class Client
...
      * @JMS\Type("string")
-     * @JMS\Groups({"edit"})
+     * @JMS\Groups({"edit", "pa-edit"})
      *
+     * Lay
      * @Assert\Length(max=200, maxMessage="client.address.maxMessage")
      *
+     * Pa
+     * @Assert\Length(max=200, maxMessage="client.address.maxMessage", groups={"pa-client"})
+     *
```
Comment (slizzio) [inline src/AppBundle/Entity/Client.php]: Multiple assertions?

File: src/AppBundle/Entity/Client.php [hunk 6]
```diff
@@ -147,23 +182,36 @@ class Client
...
      * @JMS\Type("string")
-     * @JMS\Groups({"edit"})
+     * @JMS\Groups({"edit", "pa-edit"})
      *
+     * Lay
      * @Assert\Length(min=10, max=20, minMessage="common.genericPhone.minLength", maxMessage="common.genericPhone.maxLength")
      *
+     * Pa
+     * @Assert\Length(min=10, max=20, minMessage="common.genericPhone.minLength", maxMessage="common.genericPhone.maxLength", groups={"pa-client"})
+     *
```
Comment (slizzio) [inline src/AppBundle/Entity/Client.php]: same

File: src/AppBundle/Form/Pa/ClientType.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+class ClientType extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder->add('firstname', 'text')
+                 ->add('lastname', 'text')
+                ->add('dateOfBirth', 'date', ['widget' => 'text',
+                        'input' => 'datetime',
+                        'format' => 'dd-MM-yyyy',
+                        'invalid_message' => 'Enter a valid date',
```
Comment (slizzio) [inline src/AppBundle/Form/Pa/ClientType.php]: translate?

File: src/AppBundle/Resources/views/Pa/Index/clientEdit.html.twig [hunk 8]
```diff
@@ -0,0 +1,37 @@
+{% extends 'AppBundle:Layouts:application.html.twig' %}
+
+{% block htmlTitle %}Edit client details{% endblock %}
+{% block pageTitle %}Edit client details{% endblock %}
+{% block pageTitleClass %}heading-large{% endblock %}
+
+{% block supportTitleTop %}
+    <span class="heading-secondary">Client profile: {{ client.fullname }}</span>
```
Comment (slizzio) [inline src/AppBundle/Resources/views/Pa/Index/clientEdit.html.twig]: translate

File: src/AppBundle/Resources/views/Pa/Index/clientEdit.html.twig [hunk 9]
```diff
@@ -0,0 +1,37 @@
+{% extends 'AppBundle:Layouts:application.html.twig' %}
+
+{% block htmlTitle %}Edit client details{% endblock %}
+{% block pageTitle %}Edit client details{% endblock %}
```
Comment (slizzio) [inline src/AppBundle/Resources/views/Pa/Index/clientEdit.html.twig]: translate

Comment (elvisciotti) [inline src/AppBundle/Resources/views/Pa/Index/clientEdit.html.twig]: other PA's page are not translated, so I'd prefer to have Biggs doing that and keep keys/files naming consistent. Can't do that as none of the PA is translated. We were considering not using translation at a certain point, but not sure what it was agreed as more convenient. No preferences from my side @Biggsen

File: src/AppBundle/Entity/Client.php (hunk 4 above)
Comment (elvisciotti) [inline src/AppBundle/Entity/Client.php]: yes, good point. I wanted to prefix the report-related starting with "report", but it's grammatically incorrect this way. I'll change that, including the API

File: src/AppBundle/Controller/Pa/IndexController.php (hunk 1 above)
Comment (elvisciotti) [inline src/AppBundle/Controller/Pa/IndexController.php]: in terms of ACL
- route it's under /pa, so Lay deputy cannot access it (see security.yml)
- PAs out of team won't be able to access the client given that ID, see API clientController::findByIdAction, line 98. That exception will be thrown in the controller straightaway

Comment (slizzio) [inline src/AppBundle/Controller/Pa/IndexController.php]: cool. thats fine

**Pull request merged**